line_parser.go

```diff
@@ -15,66 +15,138 @@
-				vals = strings.Split(val[1:len(val)-1], ",")
-				for i := range vals {
-					vals[i] = strings.TrimSpace(vals[i])
-				}
+	defer func() {
+		if r := recover(); r != nil {
+			if r == (parseError{}) {
+				column := len(origLine) - len(line) + 1
+				cmd = ""
+				cmdArgs = nil
+				err = errors.Errorf("cannot parse directive at column %d: %s", column, origLine)
+				// Note: to debug an unexpected parsing error, this is a good place to
+				// add a debug.PrintStack().
 			} else {
-				vals = []string{val}
+				panic(r)
 			}
 		}
-		cmdArgs = append(cmdArgs, CmdArg{Key: key, Vals: vals})
-	}
-	return cmd, cmdArgs, nil
-}
+	}()
 
-var splitDirectivesRE = regexp.MustCompile(`^ *[-a-zA-Z0-9/_,\.]+(|=[-a-zA-Z0-9_@=+/,\.]*|=\([^)]*\))( |$)`)
+	// until removes the prefix up to one of the given characters from line and
+	// returns the prefix.
+	until := func(chars string) string {
+		idx := strings.IndexAny(line, chars)
+		if idx == -1 {
+			idx = len(line)
+		}
+		res := line[:idx]
+		line = line[idx:]
+		return res
+	}
 
-// splits a directive line into tokens, where each token is
-// either:
-//  - a,list,of,things        # this is just one argument
-//  - argument
-//  - argument=a,b,c,d        # this is just one value string
-//  - argument=               # = empty value string
-//  - argument=(values, ...)  # a comma-separated array of value strings
-func splitDirectives(line string) ([]string, error) {
-	var res []string
+	cmd = until(" ")
+	if cmd == "" {
+		panic(parseError{})
+	}
+	line = strings.TrimSpace(line)
 
-	origLine := line
 	for line != "" {
-		str := splitDirectivesRE.FindString(line)
-		if len(str) == 0 {
-			column := len(origLine) - len(line) + 1
-			return nil, errors.Newf("cannot parse directive at column %d: %s", column, origLine)
+		var arg CmdArg
+		arg.Key = until(" =")
+		if arg.Key == "" {
+			panic(parseError{})
+		}
+		if line != "" && line[0] == '=' {
+			// Skip the '='.
+			line = line[1:]
+
+			if line == "" || line[0] == ' ' {
+				// Empty value.
+				arg.Vals = []string{""}
+			} else if line[0] != '(' {
+				// Single value.
+				val := until(" ")
+				arg.Vals = []string{val}
+			} else {
+				// Skip the '('.
+				line = line[1:]
+
+				// To allow for nested parens, we implement a recursive descent parser
+				// which operates on a slice of runes and keeps track of the current
+				// position.
+				var runes []rune
+				var pos int
+
+				// untilMatchedParen advances pos until a ')' is reached that does not
+				// match an earlier '('. After the call, pos will point right after ')'.
+				//
+				// For example, if runes[pos:] is "x+(a+(b+c)))y" before the call, it
+				// will be "y" after the call.
+				var untilMatchedParen func()
+				untilMatchedParen = func() {
+					for pos < len(runes) {
+						pos++
+						if runes[pos-1] == ')' {
+							// Done.
+							return
+						}
+						if runes[pos-1] == '(' {
+							untilMatchedParen()
```

Comment:
You might have explored this already, but this might be simplified by just keeping a counter the represents currently unmatched `(`. Traverse the runes, increment it when you see a `(`, decrement when you see a `)`. If the count goes negative, the parens are invalid. If the count is positive and you're at the end of the line, the parens are invalid. When the count is zero, you've finished parsing the arg.

Is there a case I'm not thinking of where this wouldn't work?